### PR TITLE
Arm subsystem

### LINKED ISSUE #18: Arm subsystem
the arm subsystem includes the 3 joints, the inverse kinematics for the positions of those joints, the various deployment states, auto stowing, upgrade ability from mid to high as a sequence

### PR DESCRIPTION
**Pull request by lmaxwell24**

### CODE REVIEW

File: commands/arm/statearmposition.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+from commands2 import CommandBase
+from wpilib import DataLogManager
+from subsystems.armsubsystem import ArmSubsystem
+
+
+class SetArmPositionMid(CommandBase):
+    state = ArmSubsystem.ArmState.Mid
+
+    def __init__(self, armSubsystem: ArmSubsystem) -> None:
```
Comment (LandonBayer) [inline commands/arm/statearmposition.py:9]: Why does the init only happen on the mid one? I see the others take it as a parameter but I would think it makes more sense to have a SetArmPosition class and then have another one for Mid specifically

File: constants.py [hunk 2]
```diff
@@ -487,6 +494,136 @@
...
+# Arm
+kArmPIDSlot = 0
+
+kElbowArmCANId = 30
+kElbowArmPGain = 0.2
+kElbowArmIGain = 0.0
+kElbowArmDGain = 0
+kElbowArmInverted = False
+
+kElbowArmGearRatio = 40 / 1
```
Comment (LandonBayer) [inline constants.py]: Is this true?

File: constants.py [hunk 3]
```diff
@@ -487,6 +494,136 @@
...
+kElbowMinAngle = Rotation2d.fromDegrees(0)
+kElbowMaxAngle = Rotation2d.fromDegrees(150)
+
+kShoulderArmCANId = 44
+kShoulderArmPGain = 0.2
+kShoulderArmIGain = 0.0
+kShoulderArmDGain = 20
+kShoulderArmInverted = False
+
+kShoulderArmGearRatio = 70 / 14
```
Comment (LandonBayer) [inline constants.py]: also wrong I think

File: physics.py [hunk 4]
```diff
@@ -110,14 +112,126 @@ def update(self, tm_diff: float, robotVoltage: float) -> None:
...
+        shoulderSimMotor: TalonFXSimCollection,
+        elbowSimMotor: TalonFXSimCollection,
+        wristSimMotor: TalonFXSimCollection,
+        shoulderSimEncoder: CANCoderSimCollection,
+        elbowSimEncoder: CANCoderSimCollection,
+        wristSimEncoder: CANCoderSimCollection,
+    ) -> None:
+        self.shoulderGearbox = DCMotor.falcon500(1)
+        self.elbowGearbox = DCMotor.falcon500(1)
+        self.wristGearbox = DCMotor.falcon500(1)
```
Comment (LandonBayer) [inline physics.py:151]: Why doesn't this use the special metaclass and all that?

File: physics.py [hunk 5]
```diff
@@ -110,14 +112,126 @@ def update(self, tm_diff: float, robotVoltage: float) -> None:
...
+        self.elbowJointSim = SingleJointedArmSim(
+            self.elbowGearbox,
+            constants.kElbowArmGearRatio,
+            SingleJointedArmSim.estimateMOI(
+                constants.kArmelbowLength, constants.kArmelbowMass
+            ),
+            constants.kArmelbowLength,
+            -180 / constants.kRadiansPerDegree,
+            180 / constants.kRadiansPerDegree,
+            constants.kArmelbowMass,
```
Comment (LandonBayer) [inline physics.py]: It is mildly annoying that Elbow isn't capitalized

File: robotcontainer.py [hunk 6]
```diff
@@ -93,6 +103,23 @@ def configureButtonBindings(self):
         and then passing it to a JoystickButton.
         """
 
+        commands2.button.JoystickButton(*self.operatorInterface.armMid).whileHeld(
```
Comment (LandonBayer) [inline robotcontainer.py]: Is this designed to have a second driver on the arm? Doesn't really matter rn though

File: subsystems/armsubsystem.py [hunk 7]
```diff
@@ -0,0 +1,632 @@
...
+        self.totalCOM = Translation2d()
+        self.elbowRelativeCOM = Translation2d()
+        self.wristRelativeCOM = Translation2d()
+
+        self.reset()
+        Preferences.initBoolean(constants.kArmSmoothKey, True)
+        Preferences.initBoolean(constants.kArmObeyEndstopsKey, True)
+
+    def reset(self) -> None:
+        shoulderEncoderPosition = self.shoulderEncoder.getPosition().radians()
```
Comment (LandonBayer) [inline subsystems/armsubsystem.py]: Why doesn't it reset to 0?

File: commands/arm/parseBVH.sh [hunk 8]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh 
+# takes in two arguments, 1 which is the BVH file and a second for the output, converts only the rotation on each linkage that we want
+tail -n +38 "$1" | awk '{ print $4,$7,$10; }' > "$2"
```
Comment (LandonBayer) [inline commands/arm/parseBVH.sh:3]: I can't read this at all so I'm going to assume it's fine

File: commands/auto/autonomousaction.py [hunk 9]
```diff
@@ -23,17 +30,20 @@ class AutonomousRoutine(SequentialCommandGroup):
...
         simultaneousCommands: List[Command],
     ):
         self.name = name
         self.markerMap = {  # later todo: actual implementation
-            "store": WaitCommand(2),
-            "top": WaitCommand(2),
-            "mid": WaitCommand(2),
-            "hybrid": WaitCommand(2),
+            "store": SetArmPositionStored(arm),
+            "top": SequentialCommandGroup(SetArmPositionTop(arm), WaitCommand(0.8)),
```
Comment (LandonBayer) [inline commands/auto/autonomousaction.py:40]: Why a wait on this one?

File: commands/light/cubeLights.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        self.setName(__class__.__name__)
+
+        self.lights = lightSubsystem
+        self.addRequirements([self.lights])
+
+    def execute(self) -> None:
+        self.lights.coneLights()
+
+    def isFinished(self) -> bool:
+        return False
```
Comment (LandonBayer) [inline commands/light/cubeLights.py:36]: false?

File: constants.py [hunk 11]
```diff
@@ -580,6 +582,202 @@
...
+kArmTopSafePosition = Pose2d(
+    -10.430 * kMetersPerInch,
+    49.622 * kMetersPerInch,
+    Rotation2d.fromDegrees(51),
+)
+kArmGroundConeIntakePosition = Pose2d(
+    -30.357 * kMetersPerInch, -4.844 * kMetersPerInch, Rotation2d.fromDegrees(-115.685)
+)
+
+kArmFudgeFactorIncremetns = 0.5 * constants.kMetersPerInch
```
Comment (LandonBayer) [inline constants.py]: Increments*

File: subsystems/armsubsystem.py [hunk 12]
```diff
@@ -0,0 +1,830 @@
...
+        self.shoulderArm.set(
+            Falcon.ControlMode.Position,
+            shoulderArmEncoderPulses,
+            self._shoulderFF(Rotation2d(trueShoulderPos)),
+        )
+        self.wristArm.set(
+            Falcon.ControlMode.Position,
+            wristArmEncoderPulses,
+            self._wristFF(Rotation2d(trueWristPos)),
+        )
```
Comment (LandonBayer) [inline subsystems/armsubsystem.py:788]: Maybe I can't read but I swear I only saw the Cartesian interpolation and one bit of Jointspace throughout it, is there one that's just better or something?

**Review by LandonBayer — DISMISSED** [review]

File: commands/arm/statearmposition.py (hunk 1 above)
Comment (lmaxwell24) [reply commands/arm/statearmposition.py:9]: that would make sense, the intention is that the others inherit from it with just the change to the state subvariable of the class

File: commands/arm/parseBVH.sh (hunk 8 above)
Comment (lmaxwell24) [reply commands/arm/parseBVH.sh:3]: cuts off the first 38 lines of the BVH file, and then takes the 4th, 7th, and 10th things in each line and puts it into the file
shell scripting magic! look up: awk

File: commands/auto/autonomousaction.py (hunk 9 above)
Comment (lmaxwell24) [reply commands/auto/autonomousaction.py:40]: give a bit of time once at the position for things to settle -- probably adding a wait to mid would be a good idea

File: commands/light/cubeLights.py (hunk 10 above)
Comment (lmaxwell24) [reply commands/light/cubeLights.py:36]: have it continuously run until interrupted

File: physics.py (hunk 4 above)
Comment (lmaxwell24) [reply physics.py:151]: this is physics.py, we are interfacing with the simulated things that would happen with the motor values, which the motor properties are defined by in falcon500

File: subsystems/armsubsystem.py (hunk 12 above)
Comment (lmaxwell24) [reply subsystems/armsubsystem.py:788]: its only those special cases in which the interpolation should exist at those stages -- relatively simple given how we swap between the two

File: constants.py (hunk 2 above)
Comment (lmaxwell24) [reply constants.py]: outdated

File: subsystems/armsubsystem.py (hunk 7 above)
Comment (lmaxwell24) [reply subsystems/armsubsystem.py]: in the initialization of cancoder, it sets the offset and it takes absolute position
the resetting is done there

File: robotcontainer.py (hunk 6 above)
Comment (lmaxwell24) [reply robotcontainer.py]: all buttons may be changed

File: commands/auto/autonomousaction.py (hunk 9 above)
Comment (LandonBayer) [reply commands/auto/autonomousaction.py:40]: Doesn't it wait until it hits the position anyway? in the command or subsystem iirc it doesn't finish until the position is achieved

File: physics.py (hunk 4 above)
Comment (LandonBayer) [reply physics.py:151]: the falcon class does simulator too no?

**Review by LandonBayer — DISMISSED** [review]

File: physics.py (hunk 4 above)
Comment (lmaxwell24) [reply physics.py:151]: this is a more physical simulation -- they both work together however the metaclass isn't designed to be used in physics simulation

File: commands/auto/autonomousaction.py (hunk 9 above)
Comment (lmaxwell24) [reply commands/auto/autonomousaction.py:40]: it first hits the position based on motor stuff, I might need to adjust the position tolerences

**Review by LandonBayer — DISMISSED** [review]
I assume something went terribly wrong in override but it's fixed now 👍

**Review by caloisio — APPROVED** [review]
Currently running on robot. Advise merge to main and then fix bugs as they appear.

**Review by LandonBayer — APPROVED** [review]
Seems fine, why take it out though?

**Comment by lmaxwell24** [comment]
covered in #40

**Pull request closed without merging**